imperative.md

```diff
@@ -0,0 +1,23 @@
+# Should KCL be imperative?
+
+Currently KCl tries to keep all its variables immutable. Should we allow mutation?
+
+## Why immutable
+
+ - Fewer chances of bugs when variables cannot be updated
+ - Encourages users to use higher-level constructs like patterns, that can be optimized in both KCL and the engine much more easily than a for loop.
+ - Easy to map a specific geometric feature (e.g. a 2D polygon) to its definition in the IDE because a geometric feature is created once and then never changed, so it's fully defined at its source range. If geometry could be updated, we'd need the IDE to show where it was defined AND where the relevant change was made.
+ - KCL is about defining geometry, not about modeling a changing system. Mutation doesn't seem necessary.
+
+## Why mutation
+
+ - Mechanical engineers (MEs) will be coming from Python or JS. They will be used to imperative programming with loops and variables. The functional programming patterns (e.g. map and reduce) they'll need to engage with will be unfamiliar.
+ - Given that we're already asking MEs to go outside their comfort zone (by coding rather than using the GUI), we should make the language as intuitive as possible. This probably means for loops.
```

Comment:
See https://github.com/KittyCAD/kcl-experiments/pull/15 for a proposal for a 'for loop' construct which subsumes for_each, map, and reduce (in most cases) and works with immutability, and I believe is easy to teach to programmers coming from Python/JS (though the actual details of how it works are somewhat complex).